### ISSUE TITLE
Ask for advice about using my int8gemm 

### ISSUE THREAD

**Issue by knjwhn** [apache/incubator-mxnet#16749]
Hello everyone.
I implemented a gemm multiply function for u8s8s32 and s8s8s32 data types , and the function interface are same as openblas , I want to use it in mxnet and I realized only after quantization can we use the int8 gemm ,but when I look into source code , I find the quantization code in convolution isn't include the  gemm but link to the normal convolution code which is for float data type. and the only concerned gemm code in quantization is in fully_connected  , in which called mkldnn api cblas_gemm_s8u8s32,However , that must use mkl or mkldnn ,and I want to avoid using them instead using my own gemm function , Can anyone give some suggestion?

**Comment by knjwhn** [comment]
https://github.com/apache/incubator-mxnet/blob/c38b52784325380f79cafb9f0407ad327554fe6b/src/operator/quantization/quantized_fully_connected.cc#L270

**Comment by TaoLv** [comment]
Are you running experiments or targeting for upstreaming in the future? If just experiments, I think you can directly replace the `cblas_gemm_s8u8s32` function call with yours. Before that, you need make sure that they have the same functionality. See the document of `cblas_gemm_s8u8s32` at https://software.intel.com/en-us/mkl-developer-reference-c-cblas-gemm-1 

If you're not building MXNet with `USE_BLAS=mkl`, remember to remove the check of `MSHADOW_USE_MKL == 1` around that function. Also avoid using things like `MKL_INT8`.

**Comment by knjwhn** [comment]
thanks !  I build MXNet with  USE_BLAS=openblas , because of my target machine is arm64 so I can't using mkl or mkldnn, and by avoiding this I created int8 gemm with the same interface as openblas, and I want to use it in mxnet,  So is there have any idea to use my own int8gemm in convolution  ?

**Comment by TaoLv** [comment]
I see. Given you already have INT8 GEMM, so I would suggest you start from the normal FP32 FullyConnected operator and replace its implementation with yours. Inputs/outputs are still FP32, in the operator there should be quantization (FP32->INT8), INT8 GEMM, de-quantization (INT8/INT32 -> FP32). With this method, you can verify what's the accuracy and performance benefit you can get with your implementation. Once everything looks good, then you can extract the quantization/de-quantization code to be a normal operator which are already done for MKL-DNN/cuDNN path.

For convolution, implement an INT8 Conv API with your INT8 GEMM and do the same experiment as above.

**Comment by knjwhn** [comment]
Thanks a lot !I will try that.and is there have some study resource of how to use quantization/dequantization code?  What's more I still have a problem is that where is the current INT8 convolution using the gemm? maybe I can replace it.

**Comment by TaoLv** [comment]
MXNet INT8 Conv doesn't call GEMM directly. It leverages the Conv APIs from MKL-DNN/cuDNN.

**Comment by pengzhao-intel** [comment]
@knjwhn  Does current int8 FC OP work for you?

Could you provide some backgrounds of your workload and target with int8 GEMM?

**Comment by pengzhao-intel** [comment]
More info in our blog:  https://medium.com/apache-mxnet/model-quantization-for-production-level-neural-network-inference-f54462ebba05

**Comment by knjwhn** [comment]
Thanks a lot! my target machine caused some problem,I haven't solve it yet. Sorry I can't test it now, and I will show the performance after it fixed.

**Comment by pengzhao-intel** [comment]
> Thanks a lot! my target machine caused some problem,I haven't solve it yet. Sorry I can't test it now, and I will show the performance after it fixed.

What's your target machine?

**Comment by knjwhn** [comment]
Ok,My target machine is ft1500a , and the linux has some error during booting process.and the error message is   CPU[1] Starting init: /etc/init exists but couldn't execute it (error -13)  and I've try some tips from google but not solved yet. anyone have ideas?

**Comment by TaoLv** [comment]
Looks like the error is not related to MXNet project and you already know how to run the experiments. So I'm closing this issue. Feel free to re-open if you meet any other problem with the integration.

**Comment by knjwhn** [comment]
> Are you running experiments or targeting for upstreaming in the future? If just experiments, I think you can directly replace the `cblas_gemm_s8u8s32` function call with yours. Before that, you need make sure that they have the same functionality. See the document of `cblas_gemm_s8u8s32` at https://software.intel.com/en-us/mkl-developer-reference-c-cblas-gemm-1
> 
> If you're not building MXNet with `USE_BLAS=mkl`, remember to remove the check of `MSHADOW_USE_MKL == 1` around that function. Also avoid using things like `MKL_INT8`.

Hello , I'm so sorry if I bother you . and I have a question here. I directly replace the `cblas_gemm_s8u8s32` function with my function, and I got the similar result when I am running example/quantization/ with the command from quantization/README.md 
`python imagenet_inference.py --symbol-file=./model/mobilenet1.0-quantized-5batches-naive-symbol.json --param-file=./model/mobilenet1.0-quantized-0000.params --rgb-mean=123.68,116.779,103.939 --rgb-std=58.393,57.12,57.375 --num-skipped-batches=50 --batch-size=64 --num-inference-batches=500 --dataset=./data/val_256_q90.rec --ctx=cpu`
and the result I got is here :
before I replace:
INFO:logger:Finished inference with 32000 images
INFO:logger:Finished with 24.355488 images per second
INFO:logger:('accuracy', 0.72)
INFO:logger:('top_k_accuracy_5', 0.90475)

after I replace:
INFO:logger:Finished inference with 32000 images
INFO:logger:Finished with 23.792868 images per second
INFO:logger:('accuracy', 0.7173125)
INFO:logger:('top_k_accuracy_5', 0.9046875)

so, I was wondering if the code really called the QuantizedFullyConnectedForwardCPU function which contains the cblas_gemm_s8u8s32 function, and i add some print message but not shows finally. and I want to know how can I use the function called QuantizedFullyConnectedForwardCPU , Hope for your help.
